lib/Dialect/Arc/Transforms/MergeIfs.cpp

```diff
@@ -0,0 +1,365 @@
+                          << rootBlock.getParentOp()->getName() << "\n");
+  do {
+    anyChanges = false;
+    sinkOps();
+    mergeIfs();
+  } while (anyChanges);
+  return success();
+}
+
+/// Return the state/memory value being written by an op.
+static Value getPointerWrittenByOp(Operation *op) {
+  if (auto write = dyn_cast<StateWriteOp>(op))
+    return write.getState();
+  if (auto write = dyn_cast<MemoryWriteOp>(op))
+    return write.getMemory();
+  return {};
+}
+
+/// Return the state/memory value being read by an op.
+static Value getPointerReadByOp(Operation *op) {
+  if (auto read = dyn_cast<StateReadOp>(op))
+    return read.getState();
+  if (auto read = dyn_cast<MemoryReadOp>(op))
+    return read.getMemory();
+  return {};
+}
+
+namespace {
+/// An integer indicating the position of an operation in its parent block. The
+/// first field is the initial order/position assigned. The second field is used
+/// to order ops that were moved to the same location, which makes them have
+/// same first field.
+using OpOrder = std::pair<unsigned, unsigned>;
+
+/// A helper that tracks an op and its order, and allows for convenient
+/// substitution with another op that has a higher/lower order.
+struct OpAndOrder {
+  Operation *op = nullptr;
+  OpOrder order = {0, 0};
+
+  explicit operator bool() const { return op; }
+
+  /// Assign `other` if its order is lower than this op, or this op is null.
+  void minimize(const OpAndOrder &other) {
+    if (!op || (other.op && other.order < order))
+      *this = other;
+  }
+
+  /// Assign `other` if its order is higher than this op, or this op is null.
+  void maximize(const OpAndOrder &other) {
+    if (!op || (other.op && other.order > order))
+      *this = other;
+  }
+};
+} // namespace
+
+/// Sink operations as close to their users as possible.
+void Sinker::sinkOps() {
+  // A numeric position assigned to ops as we encounter them. Ops at the end of
+  // the block get the lowest order number, ops at the beginning the highest.
+  DenseMap<Operation *, OpOrder> opOrder;
+  // A lookup table that indicates where ops should be inserted. This is used to
+  // maintain the original op order if multiple ops pile up before the same
+  // other op that blocks their move.
+  DenseMap<Operation *, Operation *> insertionPoints;
+  // The write ops to each state/memory pointer we've seen so far. ("Next"
+  // because we run from the end to the beginning of the block.)
+  DenseMap<Value, Operation *> nextWrite;
+  // The most recent op that has an unknown (non-read/write) side-effect.
+  Operation *nextSideEffect = nullptr;
+
+  for (auto &op : llvm::make_early_inc_range(llvm::reverse(rootBlock))) {
+    // Assign an order to this op.
+    auto order = OpOrder{opOrder.size() + 1, 0};
+    opOrder[&op] = order;
+
+    // Analyze the side effects in the op.
+    op.walk([&](Operation *subOp) {
+      if (auto ptr = getPointerWrittenByOp(subOp))
+        nextWrite[ptr] = &op;
```

Comment:
This means that write operations are not marked as general side-effect ops and operations that contain reads may be moved over them. I don't think this is a problem in practice, though, because calls aren't moved anyway.